# Laboratorio 5: VALOR Y VECTOR PROPIO, POLINOMIO CARACTERÍSTICO, DIAGONALIZACIÓN

## Integrantes:

- Aranda Huerta, Milene 
- Escriba Flores, Daniel Agustin

---

In [1]:
# Importamos las librerias necesarias 
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.pipeline import make_pipeline

# ================================
import warnings
warnings.filterwarnings('ignore')


--- 
## Parte A

In [2]:
# Leemos los datos

data = pd.read_csv("vinos.csv", sep=";")
data


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        2  
1         9.8        2  
2         9.8        2  
3         9.8        3  
4         9.4        2  
...       ...      ...  
1594     10.5        2  
1595     11.2        3  
1596     11.0        3  
1597     10.2        2  
1598     11.0        3  

[1599 rows x 12 columns]

In [3]:
# verificamos los tipos de datos
data.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

In [4]:
# Elimminamos 'quality'

data1 = data.drop("quality", axis = 1)
data1

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  
0         9.4  
1         9.8  
2         9.8  
3         9.8  
4         9.4  
...       ...  
1594     10.5  
1595     11.2  
1596     11.0  
1597     10.2  
1598     11.0  

[1599 rows x 11 columns]

In [5]:
# Estandarizamos los datos
data_std = StandardScaler().fit_transform(data1)
A = np.cov(data_std, rowvar = False)
A

array([[ 1.00062578, -0.25629118,  0.67212377,  0.11484855,  0.09376383,
        -0.15389043, -0.11325227,  0.66846534, -0.68340559,  0.18312019,
        -0.06170686],
       [-0.25629118,  1.00062578, -0.55284143,  0.00191908,  0.06133613,
        -0.0105104 ,  0.07651786,  0.02204002,  0.23508431, -0.26115001,
        -0.20241462],
       [ 0.67212377, -0.55284143,  1.00062578,  0.14366701,  0.20395046,
        -0.06101629,  0.03555526,  0.36517555, -0.54224326,  0.31296577,
         0.10997202],
       [ 0.11484855,  0.00191908,  0.14366701,  1.00062578,  0.05564433,
         0.18716605,  0.20315493,  0.3555057 , -0.08570602,  0.00553058,
         0.04210177],
       [ 0.09376383,  0.06133613,  0.20395046,  0.05564433,  1.00062578,
         0.00556563,  0.04743013,  0.20075788, -0.26519198,  0.37149281,
        -0.22127893],
       [-0.15389043, -0.0105104 , -0.06101629,  0.18716605,  0.00556563,
         1.00062578,  0.66808426, -0.02195956,  0.07042154,  0.0516899 ,
        -0.069

In [6]:
# Hallamos los autovalores y autovectores
M = np.linalg.eig(A)
Autovalores, Autovectores = M[0], M[1]
print("Autovalores: \n",Autovalores)
print("\nAutovectores: \n",Autovectores)

Autovalores: 
 [3.10107182 1.92711489 1.55151379 1.21399175 0.95989238 0.05959558
 0.18144664 0.34485779 0.42322138 0.58415655 0.66002104]

Autovectores: 
 [[ 0.48931422 -0.11050274 -0.12330157 -0.22961737 -0.08261366 -0.63969145
  -0.24952314  0.19402091 -0.17759545 -0.35022736  0.10147858]
 [-0.23858436  0.27493048 -0.44996253  0.07895978  0.21873452 -0.0023886
   0.36592473 -0.1291103  -0.07877531 -0.5337351   0.41144893]
 [ 0.46363166 -0.15179136  0.23824707 -0.07941826 -0.05857268  0.0709103
   0.62167708 -0.38144967 -0.37751558  0.10549701  0.06959338]
 [ 0.14610715  0.27208024  0.10128338 -0.37279256  0.73214429 -0.18402996
   0.09287208  0.00752295  0.29984469  0.29066341  0.04915555]
 [ 0.21224658  0.14805156 -0.09261383  0.66619476  0.2465009  -0.05306532
  -0.21767112  0.11133867 -0.35700936  0.37041337  0.30433857]
 [-0.03615752  0.51356681  0.42879287 -0.04353782 -0.15915198  0.05142086
   0.24848326  0.63540522 -0.2047805  -0.11659611 -0.01400021]
 [ 0.02357485  0.5694869

--- 
## Parte B

In [7]:
# Entrenando y escalando los datos
pca_pipe = make_pipeline(StandardScaler(),PCA())
pca_pipe.fit(data1)
modelo_pca = pca_pipe.named_steps['pca']
modelo_pca

PCA()

In [8]:
# Analizamos los componente convirtiendo en dataframe
df1 = pd.DataFrame(data=modelo_pca.explained_variance_,columns=['Autovalor'],
                  index=["λ1","λ2","λ3","λ4","λ5","λ6","λ7","λ8","λ9","λ10","λ11"])
df2 = pd.DataFrame(data=modelo_pca.explained_variance_ratio_*100,columns=['%Varianza'],
                  index=["λ1","λ2","λ3","λ4","λ5","λ6","λ7","λ8","λ9","λ10","λ11"])
df3 = pd.DataFrame(data=modelo_pca.explained_variance_ratio_.cumsum()*100,columns=['%Varianza acum.'],
                  index=["λ1","λ2","λ3","λ4","λ5","λ6","λ7","λ8","λ9","λ10","λ11"])
pd.concat([df1,df2,df3],axis=1)

Autovalor  %Varianza  %Varianza acum.
λ1    3.101072  28.173931        28.173931
λ2    1.927115  17.508270        45.682201
λ3    1.551514  14.095850        59.778051
λ4    1.213992  11.029387        70.807438
λ5    0.959892   8.720837        79.528275
λ6    0.660021   5.996439        85.524714
λ7    0.584157   5.307193        90.831906
λ8    0.423221   3.845061        94.676967
λ9    0.344858   3.133110        97.810077
λ10   0.181447   1.648483        99.458561
λ11   0.059596   0.541439       100.000000

In [9]:
# convirtiendo el array a dataframe
data2 = pd.DataFrame(data= modelo_pca.components_,columns=data1.columns,
                    index = ['CP1','CP2','CP3','CP4','CP5','CP6','CP7','CP8','CP9','CP10','CP11'])
data2

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
CP1        0.489314         -0.238584     0.463632        0.146107   0.212247   
CP2       -0.110503          0.274930    -0.151791        0.272080   0.148052   
CP3       -0.123302         -0.449963     0.238247        0.101283  -0.092614   
CP4       -0.229617          0.078960    -0.079418       -0.372793   0.666195   
CP5       -0.082614          0.218735    -0.058573        0.732144   0.246501   
CP6       -0.101479         -0.411449    -0.069593       -0.049156  -0.304339   
CP7        0.350227          0.533735    -0.105497       -0.290663  -0.370413   
CP8        0.177595          0.078775     0.377516       -0.299845   0.357009   
CP9        0.194021         -0.129110    -0.381450        0.007523   0.111339   
CP10      -0.249523          0.365925     0.621677        0.092872  -0.217671   
CP11       0.639691          0.002389    -0.070910        0.184030   0.053065   

      free sulfur dioxide  total sulfur dioxide   density        pH  \
CP1             -0.036158              0.023575  0.395353 -0.438520   
CP2              0.513567              0.569487  0.233575  0.006711   
CP3              0.428793              0.322415 -0.338871  0.057697   
CP4             -0.043538             -0.034577 -0.174500 -0.003788   
CP5             -0.159152             -0.222465  0.157077  0.267530   
CP6              0.014000             -0.136308  0.391152  0.522116   
CP7              0.116596              0.093662  0.170481  0.025138   
CP8              0.204781             -0.019036  0.239223  0.561391   
CP9              0.635405             -0.592116  0.020719 -0.167746   
CP10             0.248483             -0.370750 -0.239990 -0.010970   
CP11            -0.051421              0.068702 -0.567332  0.340711   

      sulphates   alcohol  
CP1    0.242921 -0.113232  
CP2   -0.037554 -0.386181  
CP3    0.279786  0.471673  
CP4    0.550872 -0.122181  
CP5    0.225962  0.350681  
CP6    0.381263 -0.361645  
CP7    0.447469  0.327651  
CP8   -0.374604  0.217626  
CP9   -0.058367  0.037603  
CP10   0.112320 -0.303015  
CP11   0.069555 -0.314526

--- 
## Parte C

Al analizar el porcentaje de varianza acumulada, se eligen 7 componentes principales, ya que estos tiene el 90.831906% de la varianza total. cumpliendo con la condicion de que al menos sean 90%.

In [10]:
# Reprentamos el dataframe con la cantidad de componente elegidas
data3 = data2[:7]
data3

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
CP1       0.489314         -0.238584     0.463632        0.146107   0.212247   
CP2      -0.110503          0.274930    -0.151791        0.272080   0.148052   
CP3      -0.123302         -0.449963     0.238247        0.101283  -0.092614   
CP4      -0.229617          0.078960    -0.079418       -0.372793   0.666195   
CP5      -0.082614          0.218735    -0.058573        0.732144   0.246501   
CP6      -0.101479         -0.411449    -0.069593       -0.049156  -0.304339   
CP7       0.350227          0.533735    -0.105497       -0.290663  -0.370413   

     free sulfur dioxide  total sulfur dioxide   density        pH  sulphates  \
CP1            -0.036158              0.023575  0.395353 -0.438520   0.242921   
CP2             0.513567              0.569487  0.233575  0.006711  -0.037554   
CP3             0.428793              0.322415 -0.338871  0.057697   0.279786   
CP4            -0.043538             -0.034577 -0.174500 -0.003788   0.550872   
CP5            -0.159152             -0.222465  0.157077  0.267530   0.225962   
CP6             0.014000             -0.136308  0.391152  0.522116   0.381263   
CP7             0.116596              0.093662  0.170481  0.025138   0.447469   

      alcohol  
CP1 -0.113232  
CP2 -0.386181  
CP3  0.471673  
CP4 -0.122181  
CP5  0.350681  
CP6 -0.361645  
CP7  0.327651

In [11]:
# componentes como combinación lineal de las variables originales
n=len(data3)
for i in range(n):
    print("CP{:.0f} = {:.4f} (fixed acidity) + {:.4f} (volatile acidity) + {:.4f} (citric acid) + {:.4f} (residual sugar)\
+ {:.4f} (chlorides) + {:.4f} (free sulfur dioxide) + {:.4f} (total sulfur dioxide) + {:.4f} (density)+ {:.4f} (pH)\
+{:.4f} (sulphates)+ {:.4f} (alcohol)\n".
format(i+1,data3['fixed acidity'][i],data3['volatile acidity'][i],data3['citric acid'][i], data3['residual sugar'][i],\
       data3['chlorides'][i],data3['free sulfur dioxide'][i],data3['total sulfur dioxide'][i],data3['density'][i],\
       data3['pH'][i],data3['sulphates'][i],data3['alcohol'][i]))

CP1 = 0.4893 (fixed acidity) + -0.2386 (volatile acidity) + 0.4636 (citric acid) + 0.1461 (residual sugar)+ 0.2122 (chlorides) + -0.0362 (free sulfur dioxide) + 0.0236 (total sulfur dioxide) + 0.3954 (density)+ -0.4385 (pH)+0.2429 (sulphates)+ -0.1132 (alcohol)

CP2 = -0.1105 (fixed acidity) + 0.2749 (volatile acidity) + -0.1518 (citric acid) + 0.2721 (residual sugar)+ 0.1481 (chlorides) + 0.5136 (free sulfur dioxide) + 0.5695 (total sulfur dioxide) + 0.2336 (density)+ 0.0067 (pH)+-0.0376 (sulphates)+ -0.3862 (alcohol)

CP3 = -0.1233 (fixed acidity) + -0.4500 (volatile acidity) + 0.2382 (citric acid) + 0.1013 (residual sugar)+ -0.0926 (chlorides) + 0.4288 (free sulfur dioxide) + 0.3224 (total sulfur dioxide) + -0.3389 (density)+ 0.0577 (pH)+0.2798 (sulphates)+ 0.4717 (alcohol)

CP4 = -0.2296 (fixed acidity) + 0.0790 (volatile acidity) + -0.0794 (citric acid) + -0.3728 (residual sugar)+ 0.6662 (chlorides) + -0.0435 (free sulfur dioxide) + -0.0346 (total sulfur dioxide) + -0.1745 (densit